File: Sources/PhotosStorage/ListObjects.swift
import Foundation
#if canImport(FoundationXML)
import FoundationXML
#endif

extension S3Client {

    /// Lists every object under a prefix, paging internally.
    ///
    /// §4's whole sync mechanism is one LIST on `meta/`: the response carries the
    /// ETag and size of every shard, so that single request *is* the sync plan.
    ///
    /// Keys are reported **raw**. bunny.net returns zero-byte directory markers
    /// ending in `/`, and §4 puts the responsibility for skipping them in the sync
    /// diff, not here — see `S3Object.isDirectoryMarker`.
    public func list(prefix: String = "", maxKeysPerPage: Int = 1000) -> S3ListSequence {
        S3ListSequence(client: self, prefix: prefix, maxKeysPerPage: maxKeysPerPage)
    }

    /// One page of a LIST.
    func listPage(prefix: String, continuationToken: String?,
                  maxKeys: Int) async throws -> ListObjectsPage {
        var query: [(name: String, value: String)] = [
            ("list-type", "2"),
            // LIST responses are XML, and C1 control characters are illegal in XML.
            // A single such key would corrupt the parse — and LIST *is* the sync
            // mechanism — so the server percent-encodes keys for us (§2).
            ("encoding-type", "url"),
            ("max-keys", String(maxKeys)),
        ]
        if !prefix.isEmpty {
            query.append(("prefix", prefix.precomposedStringWithCanonicalMapping))
        }
        if let continuationToken {
            query.append(("continuation-token", continuationToken))
        }

        let request = try signedRequest(
            method: "GET", path: "/\(storage.zone)", query: query, body: .empty
        )
        let response = try await send(request, expecting: [200])
        return try ListObjectsParser.parse(response.body)
    }
}

struct ListObjectsPage: Sendable {
    var objects: [S3Object]
    var nextContinuationToken: String?
    var isTruncated: Bool
}

/// An `AsyncSequence` over every object under a prefix.
///
/// Callers write `for try await object in client.list(prefix:)` and never see a
/// continuation token. A full-bucket scan for `--prune` streams rather than
/// materialising ~140 000 keys at once; `collect()` covers the small `meta/` case.
public struct S3ListSequence: AsyncSequence, Sendable {
    public typealias Element = S3Object

    let client: S3Client
    let prefix: String
    let maxKeysPerPage: Int

    public struct AsyncIterator: AsyncIteratorProtocol {
        let client: S3Client
        let prefix: String
        let maxKeysPerPage: Int
        var buffer: [S3Object] = []
        var index = 0
        var continuationToken: String?
        var exhausted = false

        public mutating func next() async throws -> S3Object? {
            while index >= buffer.count {
                if exhausted { return nil }
                let page = try await client.listPage(
                    prefix: prefix,
                    continuationToken: continuationToken,
                    maxKeys: maxKeysPerPage
                )
                buffer = page.objects
                index = 0
                continuationToken = page.nextContinuationToken
                exhausted = !page.isTruncated || page.nextContinuationToken == nil
                if buffer.isEmpty && exhausted { return nil }
            }
            defer { index += 1 }
            return buffer[index]
        }
    }

    public func makeAsyncIterator() -> AsyncIterator {
        AsyncIterator(client: client, prefix: prefix, maxKeysPerPage: maxKeysPerPage)
    }

    /// Drains the sequence into an array. Fine for `meta/` (a few hundred shards);
    /// avoid it for a full-bucket scan.
    public func collect() async throws -> [S3Object] {
        var result: [S3Object] = []
        for try await object in self { result.append(object) }
        return result
    }
}

/// Parses a `ListObjectsV2` response.
///
/// Foundation's `XMLParser` rather than a hand-rolled scanner: entity escapes and
/// odd codepoints land on exactly the data path §2 flags as fragile, and this is a
/// real XML parser. On Linux it lives in `FoundationXML`, not `Foundation`.
enum ListObjectsParser {
    static func parse(_ data: Data) throws -> ListObjectsPage {
        final class Delegate: NSObject, XMLParserDelegate {
            var objects: [S3Object] = []
            var isTruncated = false
            var nextToken: String?
            var encodingType: String?

            var inContents = false
            var element = ""
            var buffer = ""
            var key = ""
            var size: Int64 = 0
            var etag: ETag?
            var lastModified: Date?

            // Instance-scoped: ISO8601DateFormatter is not Sendable, and one
            // delegate belongs to one parse.
            let iso8601: ISO8601DateFormatter = {
                let f = ISO8601DateFormatter()
                f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                return f
            }()
            let iso8601NoFraction: ISO8601DateFormatter = {
                let f = ISO8601DateFormatter()
                f.formatOptions = [.withInternetDateTime]
                return f
            }()

            func parser(_ p: XMLParser, didStartElement e: String, namespaceURI: String?,
                        qualifiedName: String?, attributes: [String: String]) {
                element = e
                buffer = ""
                if e == "Contents" {
                    inContents = true
                    key = ""; size = 0; etag = nil; lastModified = nil
                }
            }

            func parser(_ p: XMLParser, foundCharacters s: String) { buffer += s }

            func parser(_ p: XMLParser, didEndElement e: String, namespaceURI: String?,
                        qualifiedName: String?) {
                defer { buffer = "" }
                if e == "Contents" {
                    inContents = false
                    objects.append(S3Object(key: key, size: size,
                                            etag: etag, lastModified: lastModified))
                    return
                }
                if inContents {
                    switch e {
                    case "Key": key = buffer
                    case "Size": size = Int64(buffer.trimmingCharacters(in: .whitespaces)) ?? 0
                    case "ETag": etag = ETag(header: buffer)
                    case "LastModified":
                        lastModified = iso8601.date(from: buffer)
                            ?? iso8601NoFraction.date(from: buffer)
                    default: break
                    }
                } else {
                    switch e {
                    case "IsTruncated":
                        isTruncated = buffer.trimmingCharacters(in: .whitespaces) == "true"
                    case "NextContinuationToken":
                        nextToken = buffer
                    case "EncodingType":
                        encodingType = buffer.trimmingCharacters(in: .whitespaces)
                    default: break
                    }
                }
            }
        }

        let delegate = Delegate()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        guard parser.parse() else {
            let reason = parser.parserError.map { "\($0)" } ?? "unknown XML error"
            throw StorageError.malformedResponse("could not parse the LIST response: \(reason)")
        }

        // We always ask for encoding-type=url, so keys arrive percent-encoded and
        // must be decoded exactly once. A server that ignored the parameter would
        // otherwise have its literal '%' sequences mangled — hence the echo check.
        let decode = delegate.encodingType?.lowercased() == "url"
        let objects = delegate.objects.map { object in
            guard decode else { return object }
            let decoded = object.key.removingPercentEncoding ?? object.key
            return S3Object(key: decoded, size: object.size,
                            etag: object.etag, lastModified: object.lastModified)
        }
        let nextToken = delegate.nextToken.flatMap {
            decode ? ($0.removingPercentEncoding ?? $0) : $0
        }

        return ListObjectsPage(objects: objects,
                               nextContinuationToken: nextToken,
                               isTruncated: delegate.isTruncated)
    }
}

File: domain/src/commonMain/kotlin/net/stho/photos/storage/ListObjects.kt
package net.stho.photos.storage

import io.ktor.client.statement.bodyAsBytes
import io.ktor.http.decodeURLPart
import kotlin.time.Instant
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import net.stho.photos.MalformedResponseFailure
import nl.adaptivity.xmlutil.EventType
import nl.adaptivity.xmlutil.xmlStreaming

/**
 * Every object under a prefix, paging internally.
 *
 * §4's whole sync mechanism is one LIST on `meta/`: the response carries the ETag and size of
 * every shard, so that single request *is* the sync plan.
 *
 * A `Flow` rather than a list, so a full-bucket scan for `--prune` streams rather than
 * materialising ~140 000 keys at once; `toList()` covers the small `meta/` case. The caller
 * never sees a continuation token.
 *
 * Keys are reported **raw**. bunny.net returns zero-byte directory markers ending in `/`, and
 * §4 puts the responsibility for skipping them in the sync diff, not here — see
 * [S3Object.isDirectoryMarker].
 */
public fun S3Client.list(prefix: String = "", maxKeysPerPage: Int = 1000): Flow<S3Object> = flow {
    var continuationToken: String? = null
    while (true) {
        val page = listPage(prefix, continuationToken, maxKeysPerPage)
        for (object_ in page.objects) emit(object_)
        continuationToken = page.nextContinuationToken
        // A truncated page with no token would otherwise ask for the same page forever.
        if (!page.isTruncated || continuationToken == null) return@flow
    }
}

/** One page of a LIST. */
internal suspend fun S3Client.listPage(
    prefix: String,
    continuationToken: String?,
    maxKeys: Int,
): ListObjectsPage {
    val query = buildList {
        add("list-type" to "2")
        // LIST responses are XML, and C1 control characters are illegal in XML. A single such
        // key would corrupt the parse — and LIST *is* the sync mechanism — so the server
        // percent-encodes keys for us (§2).
        add("encoding-type" to "url")
        add("max-keys" to maxKeys.toString())
        if (prefix.isNotEmpty()) add("prefix" to prefix)
        if (continuationToken != null) add("continuation-token" to continuationToken)
    }
    // LIST addresses the zone itself, not a key under it.
    val response = send("GET", "/${storage.zone}", query = query, accepting = setOf(200))
    return response.bodyAsBytes().decodeToString().asListObjectsPage()
}

internal data class ListObjectsPage(
    val objects: List<S3Object>,
    val nextContinuationToken: String?,
    val isTruncated: Boolean,
)

/**
 * Parses a `ListObjectsV2` response.
 *
 * A real XML parser rather than a hand-rolled scanner: entity escapes and odd codepoints land
 * on exactly the data path §2 flags as fragile.
 *
 * The completeness check is ours, not the parser's, and it is the most important line in this
 * file. xmlutil streams, so a document that is cut off mid-way yields the elements it did read
 * and only then throws — and a document that is empty, or whose root is never closed, may not
 * throw at all. That is the §2 failure mode in its worst form: a connection dropped mid-LIST
 * would parse as zero objects, and §4's diff reads a missing key as "album deleted". A
 * truncated response would delete the whole catalog. So a parse counts only when the
 * `ListBucketResult` element was both opened *and* closed.
 */
internal fun String.asListObjectsPage(): ListObjectsPage {
    val objects = mutableListOf<S3Object>()
    var isTruncated = false
    var nextToken: String? = null
    var encodingType: String? = null
    var sawRoot = false
    var closedRoot = false

    var inContents = false
    val buffer = StringBuilder()
    var key = ""
    var size = 0L
    var etag: ETag? = null
    var lastModified: Instant? = null

    try {
        val reader = xmlStreaming.newGenericReader(this)
        while (reader.hasNext()) {
            when (reader.next()) {
                EventType.START_ELEMENT -> {
                    buffer.clear()
                    when (reader.localName) {
                        "ListBucketResult" -> sawRoot = true
                        "Contents" -> {
                            inContents = true
                            key = ""
                            size = 0L
                            etag = null
                            lastModified = null
                        }
                    }
                }

                EventType.TEXT, EventType.CDSECT, EventType.ENTITY_REF ->
                    buffer.append(reader.text)

                EventType.END_ELEMENT -> {
                    val text = buffer.toString()
                    buffer.clear()
                    when {
                        reader.localName == "ListBucketResult" -> closedRoot = true
                        reader.localName == "Contents" -> {
                            inContents = false
                            objects.add(S3Object(key, size, etag, lastModified))
                        }
                        inContents -> when (reader.localName) {
                            "Key" -> key = text
                            "Size" -> size = text.trim().toLongOrNull() ?: 0L
                            "ETag" -> etag = text.asETag()
                            "LastModified" -> lastModified = text.asIso8601()
                        }
                        else -> when (reader.localName) {
                            "IsTruncated" -> isTruncated = text.trim() == "true"
                            "NextContinuationToken" -> nextToken = text
                            "EncodingType" -> encodingType = text.trim()
                        }
                    }
                }

                else -> Unit
            }
        }
    } catch (failure: Exception) {
        // Deliberately broad. xmlutil signals malformed input with more than one type — an
        // `XmlException` for a bad document, an `IllegalArgumentException` for an empty one —
        // and every one of them means the same thing here: this response cannot be trusted, so
        // it must not become "the bucket is empty".
        throw MalformedResponseFailure("could not parse the LIST response: ${failure.message}")
    }

    if (!sawRoot || !closedRoot) {
        throw MalformedResponseFailure(
            "LIST response was truncated — no complete <ListBucketResult> element",
        )
    }

    // We always ask for encoding-type=url, so keys arrive percent-encoded and must be decoded
    // exactly once. A server that ignored the parameter would otherwise have its literal '%'
    // sequences mangled — hence the echo check.
    if (encodingType?.lowercase() != "url") {
        return ListObjectsPage(objects, nextToken, isTruncated)
    }
    return ListObjectsPage(
        objects = objects.map { it.copy(key = it.key.percentDecoded()) },
        nextContinuationToken = nextToken?.percentDecoded(),
        isTruncated = isTruncated,
    )
}

/** A key that is not valid percent-encoding is kept as it came, never dropped. */
private fun String.percentDecoded(): String = runCatching { decodeURLPart() }.getOrDefault(this)

private fun String.asIso8601(): Instant? = runCatching { Instant.parse(trim()) }.getOrNull()
